parallel Computation of Transition Matrix

benassa-de-glassa opened the pull request:
can compute the Transition Matrix in Parallel
requires the pathos library

benassa-de-glassa committed:
parallel Computation of Transition Matrix

can compute the Transition Matrix in Parallel
requires pathos

verginer commented on pathpy/HigherOrderNetwork.py:
```diff
@@ -873,31 +916,48 @@ def getTransitionMatrix(self, includeSubPaths=True):
+        
+        edges_length= len(self.edges)
+
+        n_chunks = n_cores
```
this will raise an error, since `n_cores` is not defined in this scope, I guess you wanted to write `n_chunks = self.cores`

benassa-de-glassa replied:
yes you're right, thank you

verginer reviewed (changes requested):
Fix undefined variable `n_cores`.

verginer commented:
@benassa-de-glassa thanks for the nice contribution.  Do you have any number on the speedup and at what scale (e.g. density, number of nodes, ...) this approach increases the speed?

I have run it on a toy model and due to the overhead it actually is slowed down.

benassa-de-glassa committed:
Update HigherOrderNetwork.py

benassa-de-glassa reviewed (commented):
fixed n_cores to self.cores

verginer commented:
Does `pathos` add some features, which are not available in the standard multiprocessing library? Or would it be possible to avoid this dependency?

benassa-de-glassa commented:
@verginer It actually slows down for smaller networks. But for the Entropy Growth Rate Ratio, at some point the transition matrix of size 16k x 16k is calculated. In that case, when using 6 cores instead of 1, the time to calculate went from ~15 minutes to ~6.5 minutes

verginer commented:
To handle small and large networks it might be good to have a heuristic (e.g. number of edges) to compute it sequentially if the network is small and move to parallel for sufficiently large networks.

benassa-de-glassa commented:
pathos improves on multiprocessing, at least, that's what they say on their pypi site.
But I mostly used it because I'm somewhat familiar with it. I'm right now reading up on the standard library and it looks similiar enough to be interchangeable

benassa-de-glassa commented:
you're certainly right, one should calculate smaller matrices sequentially. It's on my todo list

benassa-de-glassa committed:
removed pathos dependency

removed pathos, so that the standard multiprocessing library can be used, thereby removing the dependency

IngoScholtes commented:
Thanks, Benedikt! 

Awesome work, that is something we were actually missing. We will see how to best integrate it without creating problems in deployment.

benassa-de-glassa commented:
Simon just made me aware, that I made a mistake while importing some files. This means, I tested the wrong files and that the way I implemented the Pool using standard multiprocessing does not work due to a pickling issue and will still require pathos.
I'm very sorry for this mistake

verginer commented:
What is the problem exactly? I am able to use the multiprocessing version using the tests in `tests/test_estimation.py` and it correctly executes with the expected result? Do you mean that the pickling does not work for large objects?

benassa-de-glassa commented:
I ran the test too, and got no error. 
Until I changed the the file in the site-packages. I believe that the import in `test_estimation.py ` uses the site-package of pathpy instead of the package in the folder which it is saved in, which is why I had not noticed until Simon pointed it out to me

[pytest.txt](https://github.com/IngoScholtes/pathpy/files/1503209/pytest.txt)

benassa-de-glassa committed:
closed the pool after usage

not closing the pool lets the processes stay alive, even after they finished their task
